### PR TITLE
Add sign to ASN.1 Integer types for RSAPublicKey

### PR DESCRIPTION
**Pull request by hug-dev**
The RSAPublicKey modulus and exponent should be parsed from signed big
endian bytes. This should fix the CI for the TPM and PKCS 11 providers.

Signed-off-by: Hugues de Valon <hugues.devalon@arm.com>

### CODE REVIEW

**Comment by hug-dev** [comment]
The `as_bytes_be` method still exists but look at the [implementation](https://docs.rs/picky-asn1/0.1.0/src/picky_asn1/wrapper.rs.html#319-331), the first byte is removed if it is zero.
The `import_and_export_public_key` test was failing because this zero was removed:
```
thread 'per_provider::normal_tests::export_public_key::tests::import_and_export_public_key' panicked at 'assertion failed: `(left == right)`
  left: `[48, 129, 137, 2, 129, 129, 0, 153, 165, 220, 135, 89, 101, 254, 229, 28, 33, 138, 247, 20, 102, 253, 217, 247, 246, 142, 107, 51, 40, 179, 149, 45, 117, 254, 236, 161, 109, 16, 81, 135, 72, 112, 132, 150, 175, 128, 173, 182, 122, 227, 214, 196, 130, 54, 239, 93, 5, 203, 185, 233, 61, 159, 156, 7, 161, 87, 48, 234, 105, 161, 108, 215, 211, 150, 168, 156, 212, 6, 63, 81, 24, 101, 72, 160, 97, 243, 142, 86, 10, 160, 122, 8, 228, 178, 252, 35, 209, 222, 228, 16, 143, 99, 143, 146, 241, 186, 187, 22, 209, 86, 141, 24, 159, 12, 146, 44, 111, 254, 183, 54, 229, 109, 28, 39, 22, 141, 173, 85, 26, 58, 9, 128, 27, 57, 131, 2, 3, 1, 0, 1]`,
 right: `[48, 129, 136, 2, 129, 128, 153, 165, 220, 135, 89, 101, 254, 229, 28, 33, 138, 247, 20, 102, 253, 217, 247, 246, 142, 107, 51, 40, 179, 149, 45, 117, 254, 236, 161, 109, 16, 81, 135, 72, 112, 132, 150, 175, 128, 173, 182, 122, 227, 214, 196, 130, 54, 239, 93, 5, 203, 185, 233, 61, 159, 156, 7, 161, 87, 48, 234, 105, 161, 108, 215, 211, 150, 168, 156, 212, 6, 63, 81, 24, 101, 72, 160, 97, 243, 142, 86, 10, 160, 122, 8, 228, 178, 252, 35, 209, 222, 228, 16, 143, 99, 143, 146, 241, 186, 187, 22, 209, 86, 141, 24, 159, 12, 146, 44, 111, 254, 183, 54, 229, 109, 28, 39, 22, 141, 173, 85, 26, 58, 9, 128, 27, 57, 131, 2, 3, 1, 0, 1]`', tests/per_provider/normal_tests/export_public_key.rs:64:9
```
(look at index 6)

**Comment by hug-dev** [comment]
We realised that:
* the `0x00` byte in front is actually needed to differentiate a positive integer which the highest significant bit is 1 from a negative value
* TPM and PKCS11 libraries should only be given positive values for modulus and exponent and I am going to add that check
* when importing a key, the modulus given as bytes to the TPM or PKCS11 libraries should not contain the `0x00` byte (`as_bytes_be` should be used)
* when exporting a key, `0x00` should be added in front of the modulus if its HSB is 1

File: src/providers/pkcs11_provider/mod.rs [hunk 1]
```diff
@@ -600,6 +600,11 @@ impl Provide for Pkcs11Provider {
 
         let public_key: RsaPublicKey = picky_asn1_der::from_bytes(&op.key_data).unwrap();
 
+        if public_key.modulus.is_negative() || public_key.public_exponent.is_negative() {
```
Comment (ionut-arm) [inline src/providers/pkcs11_provider/mod.rs]: ❤️

File: src/providers/pkcs11_provider/mod.rs [hunk 2]
```diff
@@ -722,9 +727,24 @@ impl Provide for Pkcs11Provider {
...
+                    // To produce a valid ASN.1 RSAPublicKey structure, 0x00 is put in front of the positive
+                    // integer if highest significant bit is one, to differentiate it from a negative number.
+                    if modulus[0] & 0x80 == 0x80 {
+                        modulus.insert(0, 0x00);
+                    }
+                    if public_exponent[0] & 0x80 == 0x80 {
+                        public_exponent.insert(0, 0x00);
+                    }
+
+                    let modulus = IntegerAsn1::from_signed_bytes_be(modulus);
```
Comment (ionut-arm) [inline src/providers/pkcs11_provider/mod.rs]: I find it weird that they implemented both `as_bytes_be` and `as_signed_bytes_be`, but only `from_signed_bytes_be`. Feels like a method that does the above check and insert before the conversion should exist too... ohwell

File: src/providers/pkcs11_provider/mod.rs [hunk 3]
```diff
@@ -722,9 +727,24 @@ impl Provide for Pkcs11Provider {
                     error!("Error when extracting attribute: {}.", rv);
                     Err(ResponseStatus::PsaErrorCommunicationFailure)
                 } else {
+                    let mut modulus = attrs[0].get_bytes();
+                    let mut public_exponent = attrs[1].get_bytes();
+
+                    // To produce a valid ASN.1 RSAPublicKey structure, 0x00 is put in front of the positive
+                    // integer if highest significant bit is one, to differentiate it from a negative number.
+                    if modulus[0] & 0x80 == 0x80 {
```
Comment (ionut-arm) [inline src/providers/pkcs11_provider/mod.rs]: `if modulus[0] >= 128`
But I agree that the current check is more explicit as to what it's doing

File: src/providers/pkcs11_provider/mod.rs (hunk 2 above)
Comment (hug-dev) [reply src/providers/pkcs11_provider/mod.rs]: Asked this in Devolutions/picky-rs#12

**Pull request merged**